python/ray/tune/schedulers/pb2.py

```diff
@@ -0,0 +1,341 @@
+
+    xt = optimize_acq(UCB, m, m1, fixed, num_f)
+
+    # convert back...
+    xt = xt * (np.max(base_vals, axis=0) - np.min(base_vals, axis=0)) + np.min(
+        base_vals, axis=0)
+
+    xt = xt.astype(np.float32)
+    return (xt)
+
+
+def explore(data, bounds, current, base, old, config):
+    """ This function primarily processes the data from completed trials
+        and then requests the next config from the select_config function.
+        It then adds the new trial to the dataframe, so that the reward change
+        can be computed using the new weights.
+        It returns the new point and the dataframe with the new entry.
+    """
+
+    df = data.sort_values(by='Time').reset_index(drop=True)
+
+    # group by trial ID and hyperparams. Compute change in timesteps and reward.
+    df['y'] = df.groupby(['Trial'] +
+                         [key for key in bounds.keys()])['Reward'].diff()
+    df['t_change'] = df.groupby(['Trial'] +
+                                [key for key in bounds.keys()])['Time'].diff()
+
+    # delete entries without positive change in t.
+    df = df[df['t_change'] > 0].reset_index(drop=True)
+    df['R_before'] = df.Reward - df.y
+
+    # normalize the reward change by the update size. For example if trials took diff lengths of time.
+    df['y'] = df.y / df.t_change
+    df = df[~df.y.isna()].reset_index(drop=True)
+    df = df.sort_values(by='Time').reset_index(drop=True)
+
+    # only use the last 1k datapoints, so the GP is not too slow
+    df = df.iloc[-1000:, :].reset_index(drop=True)
+
+    # we need this to know the T and Reward for the weights
+    dfnewpoint = df[df['Trial'] == str(base)]
+
+    if not dfnewpoint.empty:
+
+        # now specify the dataset for the GP
+        y = np.array(df.y.values)
+        # meta data we keep -> episodes and reward (TODO: convert to curve)
+        t_r = df[['Time', 'R_before']]  
+        h = df[[key for key in bounds.keys()]]
+        X = pd.concat([t_r, h], axis=1).values
+        newpoint = df[df['Trial'] == str(base)].iloc[-1, :][['Time', 'R_before'
+                                                             ]].values
+        new = select_config(
+            X,
+            y,
+            current,
+            newpoint,
+            bounds,
+            num_f=len(t_r.columns),
+            num=len(h.columns))
+
+        new_config = config.copy()
+        values = []
+        names = []
+        for i, col in enumerate(h.columns):
+            if type(config[col]) is int:
+                new_config[col] = int(new[i])
+                values.append(int(new[i]))
+            else:
+                new_config[col] = new[i]
+                values.append(new[i])
+
+        new_T = df[df['Trial'] == str(base)].iloc[-1, :]['Time']
+        new_Reward = df[df['Trial'] == str(base)].iloc[-1, :].Reward
+
+        lst = [[old] + [new_T] + values + [new_Reward]]
+        cols = ['Trial', 'Time'] + [key for key in bounds.keys()] + ['Reward']
```

Comment:
I haven't benchmarked this but I think this might be faster?
```suggestion
        cols = ['Trial', 'Time'] + list(bounds) + ['Reward']
```